crates/runtime/src/datasource/spiceai.rs

```diff
@@ -0,0 +1,80 @@
+#![allow(clippy::unwrap_used)]
+
+use super::DataSource;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::auth::AuthProvider;
+use arrow::record_batch::RecordBatch;
+use futures::executor::block_on;
+use futures::StreamExt;
+use spice_flight::SpiceFlightClient;
+use tokio::sync::Mutex;
+
+pub struct SpiceAI {
+    pub spice_client: Arc<Mutex<SpiceFlightClient>>,
+    pub sleep_duration: Duration,
+}
+
+impl DataSource for SpiceAI {
+    fn new(auth_provider: Box<dyn AuthProvider>) -> Self
+    where
+        Self: Sized,
+    {
+        SpiceAI {
+            spice_client: Arc::new(Mutex::new(
+                block_on(SpiceFlightClient::new(
+                    "https://flight.spiceai.io",
+                    auth_provider.get_token(),
+                ))
+                .unwrap(),
+            )),
+            sleep_duration: Duration::from_secs(10),
+        }
+    }
+
+    fn get_all_data_refresh_interval(&self, _dataset: &str) -> Option<Duration> {
+        Some(self.sleep_duration)
+    }
+
+    fn get_all_data(
+        &self,
+        dataset: &str,
+    ) -> Pin<Box<dyn Future<Output = Vec<RecordBatch>> + Send>> {
+        let client = self.spice_client.clone();
+        let dataset = dataset.to_string();
+        Box::pin(async move {
+            let flight_record_batch_stream_result = client
+                .lock()
+                .await
+                .query(format!("SELECT * FROM {dataset}").as_str())
+                .await;
+
+            let mut flight_record_batch_stream = match flight_record_batch_stream_result {
+                Ok(stream) => stream,
+                Err(error) => {
+                    tracing::error!("Failed to query with spice client: {:?}", error);
```

Comment:
Hmm, do we need to change the return type to be a Result so that we can pass back errors?

Reply:
I prefer it this way otherwise we push the handling of errors onto the caller which will then have to handle them depending on the source. So I think it's best to keep the error handling implementation specific.